File: 프로그래머스/lv2/87946. 피로도/피로도.py
from itertools import permutations

def explore_count(k, dungeon):
    count = 0
    for li in dungeon:
        min_k, con_k = li
        if min_k > k:
            break
        count += 1
        k -= con_k
    return count

def solution(k, dungeons):
    dungeons = list(permutations(dungeons))
    dungeon_count = len(dungeons)
    max_explore_count = 0
    for dungeon in dungeons:
        max_explore_count = max(max_explore_count, explore_count(k, dungeon))
        if max_explore_count == dungeon_count:
            return max_explore_count
    return max_explore_count

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    boolean[] isVisited;
    int dungeonCount;
    int[][] dungeonsArray;
    int maxExploreCount = 0;
    public int solution(int k, int[][] dungeons) {
        dungeonCount = dungeons.length;
        isVisited = new boolean[dungeonCount];
        dungeonsArray = dungeons;
        exploreDungeons(k, 0);

        return maxExploreCount;
    }

    public void exploreDungeons(int num, int depth) {
        if(dungeonCount == depth) {
            return;
        }
        for(int i = 0; i < dungeonCount; i++) {
            if(isVisited[i]) continue;
            if(num >= dungeonsArray[i][0]) {
                if(maxExploreCount < depth + 1) maxExploreCount = depth + 1;
                isVisited[i] = true;
                exploreDungeons(num - dungeonsArray[i][1], depth + 1);
                isVisited[i] = false;
            }
        }
    }
}